sympy/polys/monomials.py

```diff
@@ -50,44 +59,103 @@ def itermonomials(variables, max_degree, min_degree = 0):
+        elif not is_sequence(min_degrees):
+            raise ValueError('min_degrees is not a list')
+        else:
+            if len(min_degrees) != n:
+                raise ValueError('Argument sizes do not match')
+            if any(i < 0 for i in min_degrees):
+                raise ValueError("min_degrees can't contain negative numbers")
+        total_degree = False
+    else:
+        max_degree = max_degrees
+        if max_degree < 0:
+            raise ValueError("max_degrees can't be negative")
+        if min_degrees is None:
+            min_degree = 0
+        else:
+            if min_degrees < 0:
+                raise ValueError("min_degrees can't be negative")
+            min_degree = min_degrees
+        total_degree = True
+    if total_degree:
+        if min_degree > max_degree:
+            return
+        if not variables or max_degree == 0:
+            yield {S(1)}
+            return
+        # Force to list in case of passed tuple or other incompatible collection
+        variables = list(variables) + [S(1)]
+        if all(variable.is_commutative for variable in variables):
+            monomials_list_comm = []
+            for item in combinations_with_replacement(variables, max_degree):
+                powers = dict()
+                for variable in variables:
+                    powers[variable] = 0
+                for variable in item:
+                    if variable != 1:
+                        powers[variable] += 1
+                if max(powers.values()) >= min_degree:
+                    monomials_list_comm.append(Mul(*item))
+            for mon in set(monomials_list_comm):
+                yield mon
+        else:
+            monomials_list_non_comm = []
+            for item in product(variables, repeat=max_degree):
+                powers = dict()
+                for variable in variables:
+                    powers[variable] = 0
+                for variable in item:
+                    if variable != 1:
+                        powers[variable] += 1
+                if max(powers.values()) >= min_degree:
+                    monomials_list_non_comm.append(Mul(*item))
+            for mon in set(monomials_list_non_comm):
+                yield mon
     else:
-        monomials_list_non_comm = []
-        for item in product(variables, repeat=max_degree):
-            powers = dict()
-            for variable in variables:
-                powers[variable] = 0
-            for variable in item:
-                if variable != 1:
-                    powers[variable] += 1
-            if sum(powers.values()) >= min_degree:
-                monomials_list_non_comm.append(Mul(*item))
-        return set(monomials_list_non_comm)
+        if any(min_degrees[i] > max_degrees[i] for i in range(n)):
+            raise ValueError('min_degrees[i] must be <= max_degrees[i] for all i')
+        power_lists = []
+        for var, min_d, max_d in zip(variables,min_degrees,max_degrees):
```

Comment:
This is a minor point, but the argument list of zip would look less crowded if a space is added after the two commas.

Reply:
Thanks again, I fixed that.